Encode small varint values (<=64) in a single byte.

petermattis opened the pull request:
MVCC key encoding (EncodeBytes) causes \x00 to be encoded as
\x00\x01. Avoid this small bit of space wastage by using \x01 as the
marker for system keys.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3332)

<!-- Reviewable:end -->

petermattis committed:
Define the encoding constants in relation to each other.

petermattis committed:
Encode small varint values (<=64) in a single byte.

petermattis committed:
Adjust the key constants to not use \x00.

MVCC key encoding (EncodeBytes) causes \x00 to be encoded as
\x00\x01. Avoid this small bit of space wastage by using \x01 as the
marker for system keys.

bdarnell commented:
LGTM. 

Why 64 instead of 128 for the small keys? We have room for it in the encoding, right? Do you have plans for that space?

---

Review status: 0 of 15 files reviewed at latest revision, 6 unresolved discussions.

---

*[config/config_test.go, line 75 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3332#-K4sNJs2W-Gl_anKv83M:-K4sNJs2W-Gl_anKv83N:1031807186) ([raw file](https://github.com/cockroachdb/cockroach/blob/3abe95b247b750c1a613e2dd28b2c25031feb664/config/config_test.go#L75)):*
Why this change? Where did "\xb3" come from and can it be a constant? (ditto for keys/printer_test.go, although that one delves deeper into the key encoding details so magic literals are a little more appropriate there)

---

*[keys/constants.go, line 45 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3332#-K4sQQredwB-E1UpV4Cm:-K4sQQredwB-E1UpV4Cn:-1159437892) ([raw file](https://github.com/cockroachdb/cockroach/blob/3abe95b247b750c1a613e2dd28b2c25031feb664/keys/constants.go#L45)):*
Originally, we used prefixes of 1-3 NUL bytes for our special keys because we wanted to reserve as much of the keyspace as possible for the application (i.e. apps could use any key that didn't start with a NUL). Now that we're moving away from that, it may make sense to use distinct prefixes (localPrefix = 0x01, metaPrefix = 0x02, systemPrefix = 0x03. We might still depend on meta and system having a common prefix, but local is definitely distinct from both of them)

---

*[storage/engine/rocksdb/encoding.cc, line 26 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3332#-K4sM7Qc8BY7KGQJLsQm:-K4sM7Qc8BY7KGQJLsQn:879425251) ([raw file](https://github.com/cockroachdb/cockroach/blob/3abe95b247b750c1a613e2dd28b2c25031feb664/storage/engine/rocksdb/encoding.cc#L26)):*
Rename this to match the corresponding go constant (kBytesMarker?). 

---

*[util/encoding/encoding.go, line 311 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3332#-K4sOBZe7CkoOaUnNvAo:-K4sOBZe7CkoOaUnNvAp:-1378872388) ([raw file](https://github.com/cockroachdb/cockroach/blob/3abe95b247b750c1a613e2dd28b2c25031feb664/util/encoding/encoding.go#L311)):*
I think it would be better to gather the constants that carve up the keyspace into one place instead of threading a chain of `fooMin = barMax+1` throughout different sections of the code. 

---

*[util/encoding/encoding.go, line 482 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3332#-K4sMhyNGWgYuZYMIbjT:-K4sMhyNGWgYuZYMIbjU:-780386485) ([raw file](https://github.com/cockroachdb/cockroach/blob/3abe95b247b750c1a613e2dd28b2c25031feb664/util/encoding/encoding.go#L482)):*
It looks like the `encoding` package no longer uses the value 0x01; add a comment here that this value is reserved as a special value for use in higher-level code (assuming I'm understanding this correctly).

---

*[util/encoding/encoding_test.go, line 264 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3332#-K4sSUyYzlNjvfqI088F:-K4sSUyZkhVJIYdvo3eD:-1031059447) ([raw file](https://github.com/cockroachdb/cockroach/blob/3abe95b247b750c1a613e2dd28b2c25031feb664/util/encoding/encoding_test.go#L264)):*
It's unfortunate that the decreasing encoding for signed integers benefits small negative numbers instead of small positive ones (or being symmetric around zero). Not sure it's worth changing, though.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3332#-:-K4sRyvOqED9Xo9LKa-w:-748732369)*
<!-- Sent from Reviewable.io -->

petermattis committed:
Gather all of the encoding tags into a single const block.

petermattis commented:
I originally did 128, but switched back to 64 when I realized that we're going to want to distinguish numbers encoded with "increasing" values from numbers encoded from "decreasing" values for key decoding. Currently, we don't distinguish the tags for these encodings, but I'm planning on fixing that which will require using the upper bit of the byte. 

---

Review status: 0 of 15 files reviewed at latest revision, 6 unresolved discussions.

---

*[config/config_test.go, line 75 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3332#-K4sNJs2W-Gl_anKv83M:-K4sUAzvOrE88bl7zMr0:-226563216) ([raw file](https://github.com/cockroachdb/cockroach/blob/3abe95b247b750c1a613e2dd28b2c25031feb664/config/config_test.go#L75)):*
`f` (0x66) is now a valid key encoding prefix (part of the float tags). Took me a while to realize what is going on. `\xb3` is outside the range. I think I'll change this to `\xff` to be more explicit. 

---

*[keys/constants.go, line 45 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3332#-K4sQQredwB-E1UpV4Cm:-K4sVoh0yh1Y7ilSbukZ:1958713067) ([raw file](https://github.com/cockroachdb/cockroach/blob/3abe95b247b750c1a613e2dd28b2c25031feb664/keys/constants.go#L45)):*
Good point, let me see what's involved here.

---

*[storage/engine/rocksdb/encoding.cc, line 26 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3332#-K4sM7Qc8BY7KGQJLsQm:-K4sW56tcvUH5lVPym1c:1837110821) ([raw file](https://github.com/cockroachdb/cockroach/blob/3abe95b247b750c1a613e2dd28b2c25031feb664/storage/engine/rocksdb/encoding.cc#L26)):*
Heh, I renamed in a follow-on change, but easy enough to do here.

---

*[util/encoding/encoding.go, line 311 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3332#-K4sOBZe7CkoOaUnNvAo:-K4sWB3kasCfYlCVmnAh:-1282733891) ([raw file](https://github.com/cockroachdb/cockroach/blob/3abe95b247b750c1a613e2dd28b2c25031feb664/util/encoding/encoding.go#L311)):*
I agree. I wanted to minimize the diff originally. I'll do so as a separate commit.

---

*[util/encoding/encoding.go, line 482 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3332#-K4sMhyNGWgYuZYMIbjT:-K4sWIKwospTIyGAFW3n:234823027) ([raw file](https://github.com/cockroachdb/cockroach/blob/3abe95b247b750c1a613e2dd28b2c25031feb664/util/encoding/encoding.go#L482)):*
Yes, the intention was to avoid using 0x01 so it could be used for system prefixes. But now I realize that isn't necessary. We just have to make sure the tags `[intMin, intMax]` are not used for system keys. Changed back to 0x01.

---

*[util/encoding/encoding_test.go, line 264 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3332#-K4sSUyYzlNjvfqI088F:-K4sWdDxaIg0jjPSImaS:163761093) ([raw file](https://github.com/cockroachdb/cockroach/blob/3abe95b247b750c1a613e2dd28b2c25031feb664/util/encoding/encoding_test.go#L264)):*
I noticed this too. I don't know if this is worth fixing. The main benefit to this change is that small increasing integers as used in table ids and column ids are encoded to a single byte.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3332#-:-K4sWoAo1oNbpXzBcKuQ:-1676200147)*
<!-- Sent from Reviewable.io -->

petermattis committed:
Change localPrefix to \x01 and SystemPrefix to \x02.

This shrinks the localPrefix from 3 to 1 bytes.

petermattis commented:
Review status: 0 of 15 files reviewed at latest revision, 6 unresolved discussions.

---

*[keys/constants.go, line 45 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3332#-K4sQQredwB-E1UpV4Cm:-K4scaMWdcs82J2OB0_2:248608792) ([raw file](https://github.com/cockroachdb/cockroach/blob/3abe95b247b750c1a613e2dd28b2c25031feb664/keys/constants.go#L45)):*
This is now done. The localPrefix is `\x01` and the SystemPrefix is `\x02`. Note that the localPrefix needs to be less than the meta prefix so that local keys don't leak into normal range data.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3332)*
<!-- Sent from Reviewable.io -->

tschottdorf commented on keys/constants.go:
```diff
@@ -115,18 +117,15 @@ var (
+	SystemMax    = roachpb.Key("\x03")
 
 	// MetaPrefix is the prefix for range metadata keys. Notice that
 	// an extra null character in the prefix causes all range addressing
```
stale comment.

petermattis replied:
Fixed.

tschottdorf commented:
Reviewed 4 of 15 files at r1, 5 of 5 files at r2, 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3332)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[keys/constants.go, line 45 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3332#-K4sQQredwB-E1UpV4Cm:-K4szKx71y_BPnohEtlW:-1967078710) ([raw file](https://github.com/cockroachdb/cockroach/blob/3abe95b247b750c1a613e2dd28b2c25031feb664/keys/constants.go#L45)):*
I think we could move the SystemPrefix to `\x03` and make the MetaPrefix `\x02` which would save 2 bytes from the meta key. I'm going to investigate this in a follow-up PR.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3332)*
<!-- Sent from Reviewable.io -->
